kotlinx-coroutines-core/common/src/flow/internal/Combine.kt

```diff
@@ -0,0 +1,142 @@
+    first: Flow<T1>, second: Flow<T2>,
+    crossinline transform: suspend FlowCollector<R>.(a: T1, b: T2) -> Unit
+) {
+    coroutineScope {
+        val firstChannel = asFairChannel(first)
+        val secondChannel = asFairChannel(second)
+        var firstValue: Any? = null
+        var secondValue: Any? = null
+        var firstIsClosed = false
+        var secondIsClosed = false
+        while (!firstIsClosed || !secondIsClosed) {
+            select<Unit> {
+                onReceive(firstIsClosed, firstChannel, { firstIsClosed = true }) { value ->
+                    firstValue = value
+                    if (secondValue !== null) {
+                        transform(getNull().unbox(firstValue), getNull().unbox(secondValue) as T2)
+                    }
+                }
+
+                onReceive(secondIsClosed, secondChannel, { secondIsClosed = true }) { value ->
+                    secondValue = value
+                    if (firstValue !== null) {
+                        transform(getNull().unbox(firstValue) as T1, getNull().unbox(secondValue) as T2)
+                    }
+                }
+            }
+        }
+    }
+}
+
+@PublishedApi
+internal fun <T, R> combine(
+    vararg flows: Flow<T>,
+    arrayFactory: () -> Array<T?>,
+    transform: suspend FlowCollector<R>.(Array<T>) -> Unit
+): Flow<R> = flow {
+    coroutineScope {
+        val size = flows.size
+        val channels =
+            Array(size) { asFairChannel(flows[it]) }
+        val latestValues = arrayOfNulls<Any?>(size)
+        val isClosed = Array(size) { false }
+
+        // See flow.combine(other) for explanation.
+        while (!isClosed.all { it }) {
+            select<Unit> {
+                for (i in 0 until size) {
+                    onReceive(isClosed[i], channels[i], { isClosed[i] = true }) { value ->
+                        latestValues[i] = value
+                        if (latestValues.all { it !== null }) {
+                            val arguments = arrayFactory()
+                            for (index in 0 until size) {
+                                arguments[index] = NULL.unbox(latestValues[index])
+                            }
+                            transform(arguments as Array<T>)
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+private inline fun SelectBuilder<Unit>.onReceive(
+    isClosed: Boolean,
+    channel: ReceiveChannel<Any>,
+    crossinline onClosed: () -> Unit,
+    noinline onReceive: suspend (value: Any) -> Unit
+) {
+    if (isClosed) return
+    channel.onReceiveOrNull {
+        // TODO onReceiveOrClosed when boxing issues are fixed
+        if (it === null) onClosed()
+        else onReceive(it)
+    }
+}
+
+// Channel has any type due to onReceiveOrNull. This will be fixed after receiveOrClosed
+private fun CoroutineScope.asFairChannel(flow: Flow<*>): ReceiveChannel<Any> = produce {
```

Comment:
It can be made fuseable with preceding buffer/flowOn calls just like `asChannel` by implementing it like this:

```
asChannelFlow().buffer(0).produceImpl(this, fair=true)
```

And adding an option `fair` parameter and threading it thorough the `ChannelFlow` method does to `collectTo` function where `SendingCollector` instance is created, so that it creates a separate `FairSendingCollector` instead.
```

Reply:
It's a bit more complicated than that.
For example, `produceIn`/`produceImpl` is not enough because we handle nulls in a special way and `onReceiveOrClosed` creates a non-eliminable box, it will be working only "partially" etc.